File: Data Structures/4 Queue/1 Queue/Queue.java
/*
    This program demonstrates the implementation of a queue using a singly linked list.

    The queue follows the FIFO (First In, First Out) principle, where elements are added at the rear 
    and removed from the front. A `Node` class represents each element, and the `Queue` class 
    manages enqueue, dequeue, and print operations.

    Primary operations:
    - Enqueue: Add an element to the rear.
    - Dequeue: Remove an element from the front.
    - Print: Display the contents of the queue.

    Note:
    - This version uses a linked list instead of an array-based or built-in implementation.
    - Java provides a built-in `Queue` interface in `java.util`, which can be used instead.
*/

public class Queue {
    
    // Node class representing an element in the queue
    static class Node {
        int data;  // Data stored in the node
        Node next; // Pointer to the next node

        public Node(int data) {
            this.data = data;
            this.next = null; // Initially, no next node
        }
    }
    
    private Node front; // Pointer to the front of the queue
    private Node rear;  // Pointer to the rear of the queue

    // Constructor to initialize an empty queue
    public Queue() {
        this.front = this.rear = null;
    }

    // Add a new element to the rear of the queue
    public void enqueue(int data) {
        Node newNode = new Node(data);
        if (rear == null) { // If queue is empty, front and rear are the same
            front = rear = newNode;
        } else {
            rear.next = newNode; // Link the new node at the end
            rear = newNode; // Update rear pointer
        }
    }
    
    // Remove an element from the front of the queue
    public void dequeue() {
        if (front == null) {
            System.out.println("Queue is empty, cannot dequeue!");
            return;
        }
        front = front.next; // Move front to the next element
        if (front == null) { // If queue is now empty, reset rear
            rear = null;
        }
    }

    // Check if the queue is empty
    public boolean isEmpty() {
        return front == null;
    }

    // Print all elements in the queue
    public void print() {
        if (isEmpty()) {
            System.out.println("Queue is empty!");
            return;
        }
        Node temp = front; // Start from the front
        while (temp != null) {
            System.out.print(temp.data + " -> ");
            temp = temp.next;
        }
        System.out.println("null"); // Indicate the end of the queue
    }
        
    public static void main(String[] args) {
        Queue q = new Queue(); // Create a queue object

        // Enqueue elements
        q.enqueue(7);
        q.enqueue(78);
        q.enqueue(11);

        System.out.println("Queue contents after enqueues:");
        q.print();

        // Dequeue an element
        q.dequeue();
        System.out.println("Queue contents after dequeue:");
        q.print();
    }
}

File: Data Structures/4 Queue/1 Queue/Queue.cpp
#include <iostream>
using namespace std;

/*
    A queue is a linear data structure that follows the First In, First Out (FIFO) principle, 
    where the first element added is the first to be removed.

    Note:
        - For simplicity and convenience, the C++ Standard Library provides a queue container (`std::queue`) 
          that offers the same functionality with built-in memory management.
        - Using the library version eliminates the need for manual memory management and linked list structures.
        
    Basic operations:
        - Enqueue: Add an element to the back of the queue.
        - Dequeue: Remove the element from the front of the queue.
        - isEmpty: Check if the queue is empty.

    Use Cases:
        - Queues are widely used in process scheduling, print job management, 
          call center systems, and algorithms like Breadth-First Search (BFS).

    Advantages:
        - The queue can dynamically grow, eliminating the need for a predefined capacity.
        - Insertions and deletions are efficient, with O(1) complexity in linked list implementations.

    Cons:
        - No random access: Unlike arrays, you cannot access elements by index; you can only interact with the front and back.
        - Memory overhead: Each element requires an extra pointer, increasing memory usage compared to arrays.

    Time Complexity:
        - Enqueue: O(1) — Adding an element updates only the `back` pointer.
        - Dequeue: O(1) — Removing an element updates only the `front` pointer.
        - isEmpty: O(1) — Checking if the queue is empty only requires a pointer comparison.
        - Print: O(n) — Traversing and printing the queue requires visiting all `n` nodes.

    Space Complexity:
        - O(n) — The space required is proportional to the number of nodes in the queue.

    Implementation:
        - This program demonstrates a queue using a singly linked list.
        - Each node contains an integer `data` and a pointer to the next node.
        - The `Queue` class manages the linked list, with `front` and `back` pointers representing both ends of the queue.
        - The destructor ensures all dynamically allocated nodes are properly deleted, preventing memory leaks.
*/

class Node {
public:
    int data;    // Data stored in the node
    Node* next;  // Pointer to the next node in the queue

    // Constructor to initialize the node with a value
    Node(int x) {
        data = x;
        next = nullptr;  // Initially, no node is linked to this one
    }
};

class Queue {
public:
    Node* front;  // Pointer to the front of the queue
    Node* back;   // Pointer to the back of the queue

    // Constructor to initialize an empty queue
    Queue() {
        front = back = nullptr;
    }

    // Destructor to delete all nodes in the queue and avoid memory leaks
    ~Queue() {
        while (!isEmpty()) {
            dequeue();
        }
    }

    // Check if the queue is empty
    bool isEmpty() const {
        return front == nullptr;
    }

    // Add a new node to the queue
    void enqueue(Node* n) {
        if (isEmpty()) {
            front = back = n;  // If the queue is empty, both pointers reference the new node
        } else {
            back->next = n;  // Link the last node to the new node
            back = n;        // Update `back` to the new node
        }
    }

    // Remove the front node from the queue
    void dequeue() {
        if (isEmpty()) {
            cout << "Queue is empty, cannot dequeue!" << endl;
            return;
        }
        Node* n = front;  // Store the front node
        front = front->next;  // Move front to the next node
        if (front == nullptr) {
            back = nullptr;  // If the queue becomes empty, reset `back`
        }
        delete n;  // Free memory of the removed node
    }

    // Print all elements of the queue
    void print() const {
        if (isEmpty()) {
            cout << "Queue is empty!" << endl;
            return;
        }
        Node* ptr = front;  // Start from the front of the queue
        while (ptr != nullptr) {
            cout << ptr->data << " -> ";
            ptr = ptr->next;  // Move to the next node
        }
        cout << "NULL" << endl;
    }
};

int main() {
    Queue Q;  // Create a queue object

    // Create nodes with data to enqueue
    Node* n1 = new Node(7);
    Node* n2 = new Node(78);
    Node* n3 = new Node(11);

    // Enqueue nodes into the queue (FIFO order)
    Q.enqueue(n1);
    Q.enqueue(n2);
    Q.enqueue(n3);

    // Print the queue contents after enqueues
    cout << "Queue contents after enqueues:" << endl;
    Q.print();

    cout << endl;

    // Dequeue one element from the queue
    Q.dequeue();

    // Print the queue contents after dequeue
    cout << "Queue contents after dequeue:" << endl;
    Q.print();

    return 0;
}
